src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData = new Object[DEFAULT_CAPACITY];
+    private int size;
+
     @Override
     public void add(T value) {
-
+        for (int i = 0; i < size; i++) {
+            if (elementData.length == size) {
+                elementData = Arrays.copyOf(elementData,
+                        elementData.length + (elementData.length >> 1));
+            }
+        }
+        elementData[size] = value;
+        size += 1;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            checkLastIndex(index);
+        } else {
+            checkIndex(index);
+        }
+        if (size + 1 >= elementData.length) {
+            elementData = Arrays.copyOf(elementData,elementData.length + (elementData.length >> 1));
+        }
+        for (int i = size; i >= index; i--) {
+            elementData[i + 1] = elementData[i];
+        }
+        elementData[index] = value;
+        size = size + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int oldSize = size;
+        size += list.size();
+        while (size > elementData.length) {
+            elementData = Arrays.copyOf(elementData,elementData.length + (elementData.length >> 1));
+        }
+        for (int i = oldSize; i < size; i++) {
+            elementData[i] = list.get(i - oldSize);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Object oldValue = elementData[index];
+        size = size - 1;
+        for (int i = index; i < size; i++) {
+            elementData[i] = elementData[i + 1];
+        }
+        return (T) oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        Object oldValue = null;
+        int pos = -1;
+        for (int i = 0; i < elementData.length; i++) {
+            if (elementData[i] == element
+                    || (elementData[i] != null && elementData[i].equals(element))) {
+                oldValue = elementData[i];
+                pos = i;
+                break;
```

Comment:
```suggestion
                return remove(i);
```
і після циклу одразу можна викидувати ексепшин NoSuchElementException